problem_builder/swipe.py

```diff
@@ -0,0 +1,151 @@
+# along with this program in a file in the toplevel directory called
+# "AGPLv3".  If not, see <http://www.gnu.org/licenses/>.
+#
+
+# Imports ###########################################################
+import logging
+from random import shuffle
+
+from xblock.fields import Scope, String, Field
+from xblock.validation import ValidationMessage
+from xblockutils.resources import ResourceLoader
+
+from .mixins import StudentViewUserStateMixin
+from .questionnaire import QuestionnaireAbstractBlock
+from .sub_api import SubmittingXBlockMixin
+
+
+# Globals ###########################################################
+log = logging.getLogger(__name__)
+loader = ResourceLoader(__name__)
+
+
+# Make '_' a no-op so we can scrape strings
+def _(text):
+    return text
+
+
+# Classes ###########################################################
+class SwipeBlock(SubmittingXBlockMixin, StudentViewUserStateMixin, QuestionnaireAbstractBlock):
+    """
+    An XBlock used to ask binary-choice questions with a swiping interface
+    """
+    CATEGORY = 'pb-swipe'
+    STUDIO_LABEL = _(u"Swipeable Binary Choice Question")
+    USER_STATE_FIELDS = ['num_attempts', 'student_choice']
+
+    message = String(
+        display_name=_("Message"),
+        help=_(
+            "General feedback provided when submitting. "
+            "(This is not shown if there is a more specific feedback tip for the choice selected by the learner.)"
+        ),
+        scope=Scope.content,
+        default=""
+    )
+
+    student_choice = String(
+        # {Last input submitted by the student
+        default="",
+        scope=Scope.user_state,
+    )
+
+    correct_choice = String(
+        display_name=_("Correct Choice"),
+        help=_("Specify the value that students may select for this question to be considered correct."),
+        scope=Scope.content,
+        values_provider=QuestionnaireAbstractBlock.choice_values_provider,
+    )
+    img_url = String(
+        display_name=_("Image"),
+        help=_(
+            "Specify the URL of an image associated with this question."
+        ),
+        scope=Scope.content,
+        default=""
+    )
+    editable_fields = QuestionnaireAbstractBlock.editable_fields + ('message', 'correct_choice', 'img_url',)
+
+    def get_results(self, previous_result):
+        correct = self.correct_choice.value == previous_result['submission']
+        return {
+            'submission': submission,
+            'message': self.message_formatted,
+            'status': 'correct' if correct else 'incorrect',
+            'weight': self.weight,
+            'score': 1 if correct else 0,
+        }
+
+    def validate_field_data(self, validation, data):
+        """
+        Validate this block's field data.
+        """
+        super(SwipeBlock, self).validate_field_data(validation, data)
+
+        def add_error(msg):
+            validation.add(ValidationMessage(ValidationMessage.ERROR, msg))
+
+        if len(self.all_choice_values) != 2:
```

Comment:
You should not add an error if there are no choices defined yet. Otherwise if you add a new swipe block in studio, and try to edit its attributes before adding the choices, it will fail to save.